libzfs_core/_libzfs_core.py

```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+            if size == 0:
+                break
+            data_bytes = os.read(fd, size)
+            result = {}
+            with nvlist_out(result) as nvp:
+                ret = _lib.nvlist_unpack(data_bytes, size, nvp, 0)
+            if ret != 0:
+                raise ZFSError(ret, "Failed to unpack list data")
+            entry = result['name']
+            if entry == name:
+                break
+            else:
+                entry = None
+    finally:
+        os.close(other_fd)
+        os.close(fd)
+
+    if entry is None:
+        raise exceptions.DatasetNotFound(name)
+    result = result['properties']
+    # In most cases the source of the property is uninteresting and the
+    # value alone is sufficient.  One exception is the 'mountpoint'
+    # property the final value of which is not the same as the inherited
+    # value.
+    mountpoint = result.get('mountpoint')
+    if mountpoint is not None:
+        mountpoint_src = mountpoint['source']
+        mountpoint_val = mountpoint['value']
+        # 'source' is the name of the dataset that has 'mountpoint' set
+        # to a non-default value and from which the current dataset inherits
+        # the property.  'source' can be the current dataset if its
+        # 'mountpoint' is explicitly set.
+        # 'source' can also be a special value like '$recvd', that case
+        # is equivalent to the property being set on the current dataset.
+        if not mountpoint_src.startswith('$'):
+            mountpoint_val = mountpoint_val + name[len(mountpoint_src):]
+    else:
+        mountpoint_val = '/' + name
+    result = { k: v['value'] for k, v in result.iteritems() }
+    if result.has_key('clones'):
+        result['clones'] = [ x for x in result['clones'] ]
+    result['mountpoint'] = mountpoint_val
+    return result
+
+
+@_uncommitted(lzc_list)
+def lzc_list_children(name):
+    '''
+    List the children of the ZFS dataset.
+
+    :param str name: the name of the dataset.
+    :return: an iterator that produces the names of the children.
+    '''
+    (fd, other_fd) = lzc_list(name, {'recurse': 1, 'type': {'filesystem': None, 'volume': None}})
+
+    def _iterator():
+        try:
+            while True:
+                record_bytes = os.read(fd, _PIPE_RECORD_SIZE)
+                if not record_bytes:
+                    break
+                (size, _, err, _, _) =  struct.unpack(_PIPE_RECORD_FORMAT, record_bytes)
+                xlate.lzc_list_children_xlate_error(err, name)
+                if size == 0:
+                    break
+                data_bytes = os.read(fd, size)
+                result = {}
+                with nvlist_out(result) as nvp:
+                    ret = _lib.nvlist_unpack(data_bytes, size, nvp, 0)
+                if ret != 0:
+                    raise ZFSError(ret, "Failed to unpack list data")
+                child = result['name']
+                if child != name:
+                    yield child
+        finally:
```

Comment:
I never imagined that an iterator / generator could end up noncollectable.  There are, of course, many thing that I never imagined about Python before :smile:.
So, let me check if I have understood your suggestion (or rather, one of the options that you suggested) correctly: get the whole list from `lzc_list` in one go, then yield its elements one by one?